src/js/mqtt.js

```diff
@@ -283,13 +297,15 @@ MqttClient.prototype.disconnect = function(err) {
  */
 MqttClient.prototype.publish = function(topic, payload, options, callback) {
   callback = callback || noop;
+
   if (!Buffer.isBuffer(payload)) {
     payload = new Buffer(payload);
   }
+  var qos = this._getQoS(options && options.qos);
   try {
     var buf = this._handle._getPublish(topic, {
-      id: this._msgId++,
-      qos: (options && options.qos) || 0,
+      id: qos === 0 ? 0 : this._getNewPacketId(),
```

Comment:
```suggestion
      id: qos === 0 ? undefined : this._getNewPacketId(),
```

Reply:
Can packet id be undefined?

Reply:
At MQTT spec, the behavior needs to be unset if QoS is zero.

Reply:
```c
if (qos > 0)
	writeInt(&ptr, packetid);
```
if qos == 0, it will not set the packetid, so 0 is suitable.